routing/index_graph.cpp

```diff
@@ -11,22 +12,502 @@ IndexGraph::IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimat
+
+void IndexGraph::Build(uint32_t jointNumber) { m_jointIndex.Build(m_roadIndex, jointNumber); }
+
 void IndexGraph::Import(vector<Joint> const & joints)
 {
   m_roadIndex.Import(joints);
-  m_jointIndex.Build(m_roadIndex, joints.size());
+  Build(joints.size());
+}
+
+void IndexGraph::GetSingleFeaturePaths(RoadPoint from, RoadPoint to,
+                                       vector<RoadPoint> & singleFeaturePath)
+{
+  CHECK_EQUAL(from.GetFeatureId(), to.GetFeatureId(), ());
+  singleFeaturePath.clear();
+  if (from == to)
+  {
+    singleFeaturePath.push_back(from);
+    return;
+  }
+
+  int shift = to.GetPointId() > from.GetPointId() ? 1 : -1;
+  for (int i = from.GetPointId(); i != to.GetPointId(); i += shift)
+    singleFeaturePath.emplace_back(from.GetFeatureId(), i);
+  singleFeaturePath.emplace_back(from.GetFeatureId(), to.GetPointId());
+}
+
+void IndexGraph::GetConnectionPaths(Joint::Id from, Joint::Id to,
+                                    vector<vector<RoadPoint>> & connectionPaths)
+{
+  CHECK_NOT_EQUAL(from, Joint::kInvalidId, ());
+  CHECK_NOT_EQUAL(to, Joint::kInvalidId, ());
+
+  vector<pair<RoadPoint, RoadPoint>> connections;
+  m_jointIndex.FindPointsWithCommonFeature(from, to, connections);
+  CHECK_LESS(0, connections.size(), ());
+
+  connectionPaths.clear();
+  for (pair<RoadPoint, RoadPoint> const & c : connections)
+  {
+    vector<RoadPoint> roadPoints;
+    GetSingleFeaturePaths(c.first /* from */, c.second /* to */, roadPoints);
+    connectionPaths.push_back(move(roadPoints));
+  }
+}
+
+void IndexGraph::GetShortestConnectionPath(Joint::Id from, Joint::Id to,
+                                           vector<RoadPoint> & shortestConnectionPath)
+{
+  vector<pair<RoadPoint, RoadPoint>> connections;
+  m_jointIndex.FindPointsWithCommonFeature(from, to, connections);
+  CHECK_LESS(0, connections.size(), ());
+
+  // Note. The case when size of connections is zero is the most common case. If not,
+  // it's necessary to perform a costy calls below.
+  if (connections.size() == 1)
+  {
+    GetSingleFeaturePaths(connections[0].first /* from */, connections[0].second /* to */,
+                          shortestConnectionPath);
+    return;
+  }
+
+  double minWeight = numeric_limits<double>::max();
+  pair<RoadPoint, RoadPoint> minWeightConnection;
+  for (pair<RoadPoint, RoadPoint> const & c : connections)
+  {
+    CHECK_EQUAL(c.first.GetFeatureId(), c.second.GetFeatureId(), ());
+    RoadGeometry const & geom = GetRoad(c.first.GetFeatureId());
+    if (!geom.IsRoad())
+      continue;
+
+    double weight = m_estimator->CalcEdgesWeight(geom, c.first.GetPointId() /* from */,
+                                                 c.second.GetPointId() /* to */);
+    if (weight < minWeight)
+    {
+      minWeight = weight;
+      minWeightConnection = c;
+    }
+  }
+
+  if (minWeight == numeric_limits<double>::max())
+  {
+    MYTHROW(RootException, ("Joint from:", from, "and joint to:", to,
```

Comment:
It connections couldn't be empty but minWeight could be equal to numeric_limits<double>::max(). If |from| and |to| are connected with a pedestrian road and at the same time we're using only car road features then geom.IsRoad() returns false. Please have a look:

```c++
  double minWeight = kInfinity;
  pair<RoadPoint, RoadPoint> minWeightConnection;
  for (auto const & c : connections)
  {
    CHECK_EQUAL(c.first.GetFeatureId(), c.second.GetFeatureId(), ());
    RoadGeometry const & geom = GetRoad(c.first.GetFeatureId());
    if (!geom.IsRoad())
      continue;

    double weight = m_estimator->CalcEdgesWeight(geom, c.first.GetPointId() /* from */,
                                                 c.second.GetPointId() /* to */);
    if (weight < minWeight)
    {
      minWeight = weight;
      minWeightConnection = c;
    }
  }

  if (minWeight == kInfinity)
  {
    MYTHROW(RoutingException, ("Joint from:", from, "and joint to:", to,
                               "are not connected a feature necessary type."));
  }
```